org.eclipse.jdt.core/model/org/eclipse/jdt/internal/core/JavaModelManager.java

```diff
@@ -999,6 +999,14 @@ public static IJavaElement create(IFolder folder, IJavaProject project) {
 			project = JavaCore.create(folder.getProject());
 			element = determineIfOnClasspath(folder, project);
 			if (element == null) {
+				IJavaProject lastProject = lastProjectUsed;
+				if (lastProject != null) {
+					// try to avoid searching through all projects
+					element = determineIfOnClasspath(folder, lastProject);
```

Comment:
Not sure this is valid to use previously used project that could correspond to already closed/deleted resource project and be not present in JavaModel anymore

Reply:
So what should happen is something like below, I haven't checked it for performance:

`lastProject = JavaModel.getJavaProject(lastProjectUsed.getElementName())`

Reply:
I have extended the Junit test to show that it works without such workaround (as the IJavaProject is a Handle only)

Reply:
This is a negative test only. What I'm more concerned about is the positive test - project exists, folder exists, we remember P1. Now P1 is closed (deleted), we still "see" P1 in lastProject and answer something weird.

Reply:
Beside this, I don't like an idea to keep a static reference to a Java Project, as a potential for a memory leak because GC will not release this project (and resource project and all the references that might be connected to them recursive) anymore.

Reply:
> Now P1 is closed

The test does it and even opens the project afterwards again. I don't get what you miss. I don't see a problem with a kept project, as it is only a handle, but i could use some weaker reference.

Reply:
No. The test checks if a non existing folder in project P2 does not exist in a project P2. 
Fine, it does not.

What I miss is that we have first a positive check  (folder F1 exists in project P1), close the project and try to create again.

Regarding some weaker reference: I assume just remembering `getElementName()` for the last project hit would do what you want.

Reply:
ok, using the String  name now